---

SQL Project
Author: Atajan Abdyyev
Date: 1/8/2021

# Project Description: 
    We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

### Description of the data

`**books**`

Contains data on books:

- `book_id`
- `author_id`
- `title`
- `num_pages` — number of pages
- `publication_date`
- `publisher_id`

`**authors**`

Contains data on authors:

- `author_id`
- `author`

`**publishers**`

Contains data on publishers:

- `publisher_id`
- `publisher`

`**ratings**`

Contains data on user ratings:

- `rating_id`
- `book_id`
- `username` — the name of the user who rated the book
- `rating`

`**reviews**`

Contains data on customer reviews:

- `review_id`
- `book_id`
- `username` — the name of the user who reviewed the book
- `text` — the text of the review

### Task

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

<div class="alert alert-success">
<b>Reviewer's comment:</b> Nice introduction!
</div>

In [1]:
# import libraries and connect to the database
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Study the tables (print the first rows).

In [2]:
#trending_by_time = pd.read_csv("trending_by_time.csv")
query = ''' SELECT * from books ''' 

books=pd.io.sql.read_sql(query, con = engine) 
display('books table head')
display(books.head())


query = ''' SELECT * from authors ''' 

authors=pd.io.sql.read_sql(query, con = engine) 
display('Authors table head')
display(authors.head())
display('******---**--'*3)

query = ''' SELECT * from publishers ''' 

publishers=pd.io.sql.read_sql(query, con = engine) 
display('Publishers table head')
display(publishers.head())
display('******---**--'*3)

query = ''' SELECT * from ratings ''' 

ratings=pd.io.sql.read_sql(query, con = engine) 
display('Ratings table head')
display(ratings.head())
display('******---**--'*3)

query = ''' SELECT * from reviews ''' 

reviews=pd.io.sql.read_sql(query, con = engine) 
display('Reviews table head')
display(reviews.head())




'books table head'

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

'Authors table head'

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

'******---**--******---**--******---**--'

'Publishers table head'

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

'******---**--******---**--******---**--'

'Ratings table head'

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

'******---**--******---**--******---**--'

'Reviews table head'

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success">
<b>Reviewer's comment:</b> Well done, let's move on to the tasks part.
</div>

In [3]:
#Find the number of books released after January 1, 2000.
query = ''' SELECT Count(book_id) as num_of_books_after_Jan1_2000 from books where publication_date > '2000-01-01' ''' 
books=pd.io.sql.read_sql(query, con = engine) 
books


num_of_books_after_jan1_2000
0                           819

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

<b> Result: There are 819  books released after January 1, 2000. </b>

In [4]:
#Find the number of user reviews and the average rating for each book. 
query = ''' SELECT books.title as book_title, Count(rating) as count_rating, avg(rating) as average_rating
    FROM ratings inner join books on books.book_id = ratings.book_id
    GROUP BY book_title
    order by average_rating desc''' 
reviews_rating=pd.io.sql.read_sql(query, con = engine) 
reviews_rating

book_title  count_rating  \
0    Captivating: Unveiling the Mystery of a Woman'...             2   
1                                        Evening Class             2   
2    In the Hand of the Goddess (Song of the Liones...             3   
3                   The Big Bad Wolf  (Alex Cross  #9)             2   
4                        A Dirty Job (Grim Reaper  #1)             4   
..                                                 ...           ...   
994  The World Is Flat: A Brief History of the Twen...             4   
995                  His Excellency: George Washington             2   
996                                              Junky             2   
997                                      Drowning Ruth             3   
998                               Harvesting the Heart             2   

     average_rating  
0              5.00  
1              5.00  
2              5.00  
3              5.00  
4              5.00  
..              ...  
994            2.25  
995            2.00  
996            2.00  
997            2.00  
998            1.50  

[999 rows x 3 columns]

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

<b> Result: Table above shows result of all books, their ratings, number of ratings. Ordered by average rating in descending order. </b>

In [5]:
#Identify the publisher that has released the greatest number of books with more than 50 pages 
#(this will help you exclude brochures and similar publications from your analysis).
#  
query = ''' SELECT  count(books.book_id) as number_of_books, publishers.publisher as publisher 
from publishers inner join books on publishers.publisher_id =books.publisher_id
 where books.num_pages > 50 
group by publisher
order by number_of_books desc 
LIMIT 1''' 
reviews_rating=pd.io.sql.read_sql(query, con = engine) 
reviews_rating

number_of_books      publisher
0               42  Penguin Books

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

<b> Result: the publisher that has released the greatest number of books with more than 50 pages is Penguin Books, 42 books. </b>

In [6]:
#Identify the author with the highest average book rating (look only at books with at least 50 ratings)
query = ''' SELECT authors.author as author, avg(rating) as average_rating,count(ratings.rating) as ratings
    FROM ((ratings inner join books on books.book_id = ratings.book_id) 
    inner join authors on books.author_id = authors.author_id)   
    GROUP BY author
    having count(ratings.rating) >50
    order by average_rating desc
    Limit 1''' 
reviews_rating=pd.io.sql.read_sql(query, con = engine) 
reviews_rating

author  average_rating  ratings
0  J.K. Rowling/Mary GrandPré        4.288462      312

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

<b> Result: J.K. Rowling/Mary GrandPré is the top author with the highest average book rating  of ~4.29, total of 312 ratings  (Hint: increase code Limit to see more authors in descending order) </b>

In [7]:
#Find the average number of text reviews among users who rated more than 50 books.

query = '''SELECT count(result.user_name) as number_of_users, AVG(result.review_count) AS average_text
            FROM (SELECT ratings.username AS user_name, COUNT(DISTINCT reviews.review_id) AS review_count
                FROM ratings
                INNER JOIN reviews ON reviews.username = ratings.username
                GROUP BY ratings.username
                HAVING COUNT(DISTINCT ratings.book_id) > 50) AS result''' 
reviews_rating=pd.io.sql.read_sql(query, con = engine) 
reviews_rating

number_of_users  average_text
0                6     24.333333

<div class="alert alert-success">
<b>Reviewer's comment: </b> ✅
</div>

<b> Result: There are 6 users who rated more than 50 books and average of text reviews among these users makes up 24 texts </b>

# Result

<b> 1: There are 819  books released after January 1, 2000. </b> <br>
<b> 2: Head of the result for the number of user reviews and the average rating for each book. (see the cell for more details) <br><br>
0	Captivating: Unveiling the Mystery of a Woman'...	2	5.00<br>
1	Evening Class	2	5.00<br>
2	In the Hand of the Goddess (Song of the Liones...	3	5.00<br>
3	The Big Bad Wolf (Alex Cross #9)	2	5.00<br>
4	A Dirty Job (Grim Reaper #1)	4	5.00<br>
...	...	...	...<br>
994	The World Is Flat: A Brief History of the Twen...	</b> <br><br>
<b> 3: the publisher that has released the greatest number of books with more than 50 pages is Penguin Books, 42 books. </b><br>
<b> 4: <b> Result: J.K. Rowling/Mary GrandPré is the top author with the highest average book rating  of ~4.29, total of 312 ratings  (Hint: increase code Limit to see more authors in descending order) </b> </b><br>
<b> 5: There are 6 users who rated more than 50 books and average of text reviews among these users makes up 24 texts </b>